### PR TITLE
CancelGeneration apiserver piece

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

- Add CancelGeneration to apiserver for modelgeneration.
- Unit tests for api/modelgeneration added for  CancelGeneration.
- Updated cmd and api for CancelGeneration to take model tags.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=generations
2. juju bootstrap localhost
2052.  juju deploy percona-cluster mysql -n 2
2056.  juju add-generation
2057.  juju cancel-generation  <-- should succeed
2056.  juju add-generation
2063.  juju advance mysql/0
2057.  juju cancel-generation  <-- should fail and list mysql/1 as the reason.
3. juju switch-generation current
2057.  juju cancel-generation  <-- should fail
4. run `juju-db.bash`
2. db.generations.find(), should report 2 docs with: active = false, completed with a non zero NumberLong.

## Documentation changes

In the future

### CODE REVIEW

File: state/modelgeneration.go [hunk 1]
```diff
@@ -221,6 +221,16 @@ func assignGenerationUnitTxnOps(id, appName, unitName string) []txn.Op {
...
+	err := g.complete(false)
+	return errors.Trace(err)
+}
+
+func (g *Generation) Cancel() error {
+	err := g.complete(true)
+	return errors.Trace(err)
+}
+
+func (g *Generation) complete(allowEmpty bool) error {
```
Comment (hmlanigan) [inline state/modelgeneration.go]: Doing it this way, we won't know if the generation was completed or canceled.  Does it matter?

Comment (hmlanigan) [reply state/modelgeneration.go]: It doesn't matter per conversations.  Though changes once we handle history are possible.

File: api/modelgeneration/modelgeneration.go [hunk 2]
```diff
@@ -40,17 +40,21 @@ func (c *Client) AddGeneration(model names.ModelTag) error {
...
 	}
-	return result.Error
+	if result.Error != nil {
+		return errors.Trace(result.Error)
+	}
+	return nil
 }
 
-// SwitchGeneration adds a model generation to the config.
+// SwitchGeneration switches a model generation to the config.
```
Comment (hmlanigan) [inline api/modelgeneration/modelgeneration.go]: a few comment updates as a drive by

File: cmd/juju/model/cancelgeneration.go [hunk 3]
```diff
@@ -94,10 +95,15 @@ func (c *cancelGenerationCommand) Run(ctx *cmd.Context) error {
...
+	if err != nil {
+		return errors.Annotate(err, "getting model details")
+	}
+
+	modelTag := names.NewModelTag(modelDetails.ModelUUID)
+	if err = client.CancelGeneration(modelTag); err != nil {
+		return err
+	}
 
 	ctx.Stdout.Write([]byte("changes dropped and target generation set to current\n"))
```
Comment (manadart) [inline cmd/juju/model/cancelgeneration.go]: I don't think we should refer to the changes as "dropped" as the next generation changes will be applied (according to the spec). Perhaps there should be a y/n confirmation with this detail.

Comment (manadart) [reply cmd/juju/model/cancelgeneration.go]: We should also give some feedback when a generation is auto-completed, but that can come in a future patch.

Comment (hmlanigan) [reply cmd/juju/model/cancelgeneration.go]: Agreed, dropped came from the spec.  @mitechie: preferences with rewording?  @manadart: https://trello.com/c/KPexZiQJ/998-add-succeeded-message-to-advance-command-when-auto-complete-happens

File: apiserver/facades/client/modelgeneration/interface.go [hunk 4]
```diff
@@ -37,6 +37,8 @@ type Generation interface {
 	Active() bool
 	AssignAllUnits(string) error
 	AssignUnit(string) error
+	CanCancel() (bool, []string, error)
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/modelgeneration/interface.go]: For some reason while implementing this, I kept getting wrapped around with the bool for canMakeCurrent and found it easier to have a helper method.  I'm open to suggestions for a different name.

File: state/modelgeneration.go (hunk 1 above)
Comment (hmlanigan) [reply state/modelgeneration.go]: added a TODO

**Comment by hmlanigan** [comment]
@manadart, correct on the qa notes, updating.

File: apiserver/facades/client/modelgeneration/interface.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply apiserver/facades/client/modelgeneration/interface.go]: CanCancelRemaining()?

File: cmd/juju/model/cancelgeneration.go (hunk 3 above)
Comment (hmlanigan) [reply cmd/juju/model/cancelgeneration.go]: "remaining incomplete changes dropped and... "

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**